### PR TITLE
Support JDBC catalog in Iceberg connector

### LINKED ISSUE #9968: Feature request: Support for jdbc catalog in Iceberg Connector
Would be great if trino supported jdbc catalog in iceberg connector

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Support JDBC catalog in Iceberg connector. 
Fixes #9968

## Release notes

(x) Release notes entries required with the following suggested text:

```markdown
# Iceberg
* Add support for JDBC catalog. ({issue}`9968`)
```

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/JdbcIcebergClient.java
Comment (meetsitaram) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/JdbcIcebergClient.java:99]: can `org.apache.iceberg.jdbc.JdbcCatalog::createNamespace` be used there instead of directly accessing the db table?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/JdbcIcebergClient.java
Comment (meetsitaram) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/JdbcIcebergClient.java:112]: can `org.apache.iceberg.jdbc.JdbcCatalog::dropNamespace` be used directly?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/JdbcIcebergClient.java
Comment (meetsitaram) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/JdbcIcebergClient.java:125]: can `org.apache.iceberg.jdbc.JdbcCatalog::listTables` be used directly?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/JdbcIcebergClient.java
Comment (meetsitaram) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/JdbcIcebergClient.java:138]: can `org.apache.iceberg.jdbc.JdbcCatalog:: newCreateTableTransaction` be used directly?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcCatalogModule.java
Comment (meetsitaram) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcCatalogModule.java:35]: Can `org.apache.iceberg.jdbc.JdbcTableOperations` be directly used instead of created a separate `JdbcIcebergTableOperations` class here?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/jdbc/TestIcebergJdbcIcebergConfig.java
Comment (meetsitaram) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/jdbc/TestIcebergJdbcIcebergConfig.java:42]: dependency on localhost postgresql can be avoided by using in-memory `h2database` for testing. this can be simply set as `jdbc:h2:mem:myH2DB`

Comment (ebyhr) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/jdbc/TestIcebergJdbcIcebergConfig.java:42]: Avoided H2 and also SQLite intentionally to make the test environment close to the actual environment.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/JdbcIcebergClient.java
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/JdbcIcebergClient.java:45]: `JdbcCatalog.listNamespaces` wouldn't work if my understanding is correct. We want to list all namespaces without providing the name, but the method expects the argument. If we call `listNamespaces()` instead of `listNamespaces(Namespace namespace)`, it fails with schema existence check.

**Comment by ebyhr** [comment]
@sungwy The root cause relates to #17. We prefer small code when adding a new enhancement, so I won't handle case insensitive matching in this PR.

**Comment by findepi** [comment]
Do we have / want to have some compatibility tests with Spark?

**Comment by meetsitaram** [comment]
Can you provide some details on how to configure s3 access with jdbc catalog? For hive catalog, we add below config in iceberg.properties file. 
```
hive.s3.endpoint=S3_ENDPOINT
hive.s3.aws-access-key=S3_ACCESS_KEY
hive.s3.aws-secret-key=S3_SECRET_KEY
```
Wondering if similar config is required with jdbc as well or some alternative exists that can access `~/.aws/config` and `~/.aws/credentials` files.

**Comment by samredai** [comment]
> Can you provide some details on how to configure s3 access with jdbc catalog? For hive catalog, we add below config in iceberg.properties file.
> 
> ```
> hive.s3.endpoint=S3_ENDPOINT
> hive.s3.aws-access-key=S3_ACCESS_KEY
> hive.s3.aws-secret-key=S3_SECRET_KEY
> ```
> 
> Wondering if similar config is required with jdbc as well or some alternative exists that can access `~/.aws/config` and `~/.aws/credentials` files.

My understanding is that it would be the same config items you've listed (and testing with the `IcebergQueryRunner` seems to confirm this). The Iceberg catalog implementations just eventually pass these to the `TrinoS3ConfigurationInitializer`. Admittedly, the `hive` prefix for the parameter names feels odd when you're using any of the other catalog types (such as JDBC here) but IMO changing that probably requires too much to bundle with this PR.

**Comment by meetsitaram** [comment]
> > Can you provide some details on how to configure s3 access with jdbc catalog? For hive catalog, we add below config in iceberg.properties file.
> > ```
> > hive.s3.endpoint=S3_ENDPOINT
> > hive.s3.aws-access-key=S3_ACCESS_KEY
> > hive.s3.aws-secret-key=S3_SECRET_KEY
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Wondering if similar config is required with jdbc as well or some alternative exists that can access `~/.aws/config` and `~/.aws/credentials` files.
> 
> My understanding is that it would be the same config items you've listed (and testing with the `IcebergQueryRunner` seems to confirm this). The Iceberg catalog implementations just eventually pass these to the `TrinoS3ConfigurationInitializer`. Admittedly, the `hive` prefix for the parameter names feels odd when you're using any of the other catalog types (such as JDBC here) but IMO changing that probably requires too much to bundle with this PR.

the same configs are working with jdbc catalog as well. thx.

**Comment by ebyhr** [comment]
Added support for product tests with Spark.

File: testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/environment/singlenode-spark-iceberg-jdbc-catalog/create-table.sql
Comment (sungwy) [inline testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/environment/singlenode-spark-iceberg-jdbc-catalog/create-table.sql:8]: It looks like the master merged iceberg PR has updated these table column names to avoid conflicts with reserved SQL keywords:
Posting it here for reference:
https://github.com/apache/iceberg/pull/3275/files#diff-1daee532e0ff506f3e2fa8186deb993eec463498c6aea8c814dcd604891aa1e1R88

Comment (ebyhr) [reply testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/environment/singlenode-spark-iceberg-jdbc-catalog/create-table.sql:8]: The column name looks still `key` https://github.com/apache/iceberg/pull/3275/files#diff-1daee532e0ff506f3e2fa8186deb993eec463498c6aea8c814dcd604891aa1e1R80. Did you post wrong URL?

File: plugin/trino-iceberg/src/test/resources/jdbc/create-table.sql
Comment (zhangminglei) [inline plugin/trino-iceberg/src/test/resources/jdbc/create-table.sql:10]: namespace here means `schema` or `database` ?

File: plugin/trino-iceberg/src/test/resources/jdbc/create-table.sql
Comment (findepi) [inline plugin/trino-iceberg/src/test/resources/jdbc/create-table.sql:3]: How do we make sure this file stays in sync with the referenced one?

or, maybe we can avoid having a copy here?

Comment (syun64) [reply plugin/trino-iceberg/src/test/resources/jdbc/create-table.sql:3]: It looks like the namespace_properties_table changes along with others we were waiting for just got released on 0.14.0 a few days ago.

Maybe we could now reference the static variable [JdbcUtil.CREATE_NAMESPACE_PROPERTIES_TABLE](https://github.com/apache/iceberg/blob/6d2edd6284ebc5301dbe45376a31ca8316852a77/core/src/main/java/org/apache/iceberg/jdbc/JdbcUtil.java#L86) here going forward

Comment (findepi) [reply plugin/trino-iceberg/src/test/resources/jdbc/create-table.sql:3]: > It looks like the namespace_properties_table changes along with others

what are the changes?
does it mean we should test this catalog against Iceberg 13 and Iceberg 14 as well?

Comment (ebyhr) [reply plugin/trino-iceberg/src/test/resources/jdbc/create-table.sql:3]: They changed column names in https://github.com/apache/iceberg/pull/5017. The PR doesn't contain logic to handle the old table definition, so I suppose they just broke the backward compatibility. 

> does it mean we should test this catalog against Iceberg 13 and Iceberg 14 as well?

I think so. (or clarify the supported version in documentation)

**Comment by ebyhr** [comment]
Just rebased on upstream to resolve logical conflicts.

File: plugin/trino-iceberg/src/test/resources/jdbc/create-table.sql
Comment (findepi) [reply plugin/trino-iceberg/src/test/resources/jdbc/create-table.sql:3]: > They changed column names in [apache/iceberg#5017](https://github.com/apache/iceberg/pull/5017). The PR doesn't contain logic to handle the old table definition, so I suppose they just broke the backward compatibility.

Concerning

**Comment by findepi** [comment]
because of https://github.com/trinodb/trino/pull/11772#discussion_r926190417 -- i am not convined we should be adding this. I don't want to be in a situation where we have a feature, but need to tell users to use something else.

**Comment by nfcampos** [comment]
> because of [#11772 (comment)](https://github.com/trinodb/trino/pull/11772#discussion_r926190417) -- i am not convined we should be adding this. I don't want to be in a situation where we have a feature, but need to tell users to use something else.

Is it possible you guys are getting confused and the previous version of this was not released? Thus this was not a breaking change? At least that's what the Iceberg folks seem to think on this thread https://github.com/apache/iceberg/issues/4952

File: plugin/trino-iceberg/src/test/resources/jdbc/create-table.sql
Comment (nfcampos) [reply plugin/trino-iceberg/src/test/resources/jdbc/create-table.sql:3]: According to this issue on Iceberg repo no backwards compatible was broken because the previous version of this was never released? https://github.com/apache/iceberg/issues/4952

Comment (nfcampos) [reply plugin/trino-iceberg/src/test/resources/jdbc/create-table.sql:3]: In fact I can confirm the Iceberg guys didn't release a breaking change, see the file in question at 0.13.2 https://github.com/apache/iceberg/blob/apache-iceberg-0.13.2/core/src/main/java/org/apache/iceberg/jdbc/JdbcUtil.java where this `NAMESPACE_PROPERTY_KEY` didn't exist, and now see same file at 0.14.0, where it does exist with the new value https://github.com/apache/iceberg/blob/apache-iceberg-0.14.0/core/src/main/java/org/apache/iceberg/jdbc/JdbcUtil.java#L83.

So sounds like it was never released with the old value

Comment (ebyhr) [reply plugin/trino-iceberg/src/test/resources/jdbc/create-table.sql:3]: @nfcampos Your confirmation looks correct. Do you know if they have compatibility test in the repository?

Comment (nfcampos) [reply plugin/trino-iceberg/src/test/resources/jdbc/create-table.sql:3]: Not that I know of, no

**Comment by syun64** [comment]
> Is it possible you guys are getting confused and the previous version of this was not released? Thus this was not a breaking change? At least that's what the Iceberg folks seem to think on this thread [apache/iceberg#4952](https://github.com/apache/iceberg/issues/4952)

This is correct. Support for iceberg_namespace_properties table was only included in 0.14.0 release.

Looking at the corresponding PRs that implement support for iceberg_namespace_properties, if we use iceberg 0.14.0 as the baseline for Trino's iceberg-jdbc catalog, I believe the following two statements hold true:
1. It will require that iceberg_namespace_properties table exists in the specified jdbc connected database
2. If 1 holds true, even iceberg tables that were created using 0.13.0 (i.e. have no entries in iceberg_namespace_properties) will be read, as these PRs were implemented with backward compatibility in mind.

[Support for Namespace properties in JDBC Catalog #3275](https://github.com/apache/iceberg/pull/3275)
[Core: Update JDBC catalog namespaces #4220](https://github.com/apache/iceberg/pull/4220)

**Comment by bitsondatadev** [comment]
:wave: @ebyhr - this PR has become inactive. If you're still interested in working on it, please let us know.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

**Comment by ebyhr** [comment]
Rebased on upstream to resolve conflicts.

**Comment by jebnix** [comment]
@bitsondatadev please don't close this PR. Deploying Trino today is super hard when using connectors that requires Hive Metastore. Hive Metastore doesn't have an official image and making the Trino deployment much harder (even though this is the smallest service).
An Iceberg JDBC connector will bring the Trino deployment to another level. To emphasize how hard and almost impossible it is to deploy production ready Hive Metastore in the k8s era, you can see that the https://github.com/trinodb/trino/issues/9968 issue is the 4th most liked issue on Trino, even more requested than the tracking issue of the Iceberg connector itself (https://github.com/trinodb/trino/issues/1324)!
@ebyhr Thank you for your work on this PR, we really need this

**Comment by bitsondatadev** [comment]
Thanks for the heads up! Keeping this open! :)

**Comment by jebnix** [comment]
@ebyhr What is left on the PR so it will get merged?

**Comment by jebnix** [comment]
@ebyhr @bitsondatadev I want to help with this. Can you clarify what's left and how can I help?

**Comment by ebyhr** [comment]
> Can you clarify what's left and how can I help?

@jebnix This PR needs review from maintainers. 
cc: @electrum @phd3

**Comment by cccs-eric** [comment]
We are also looking forward for this PR to be merged 🙏🏻 .  It is the only missing piece for us to adopt the JDBC catalog.

**Comment by jebnix** [comment]
@ebyhr I don't think I would be able to explain to you how this PR is going to be dramatic for us. In my organization, this is the only reason we can't adopt Trino. Currently even [Presto supports Iceberg without Hive Metastore](https://prestodb.io/docs/current/connector/iceberg.html#nessie-catalog). @electrum @phd3 Any update would be much appreciated!

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/JdbcIcebergClient.java
Comment (code-magician323) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/JdbcIcebergClient.java:45]: @ebyhr That PR has been merged. Isn't this let us use `listNamespaces()`?

**Comment by jebnix** [comment]
@ebyhr Can you give us an update about the progress here? This is REALLY needed

**Comment by ebyhr** [comment]
@jebnix Nothing is changed from the last comment.

**Comment by code-magician323** [comment]
@bitsondatadev @electrum @phd3 Why doesn't this get merged?

**Comment by phd3** [comment]
Thanks @ebyhr @bitsondatadev, I'll start reviewing this as well.

**Comment by bitsondatadev** [comment]
Keeping everyone in the loop. @findepi, @ebyhr, @phd3, and I will meet next week to discuss if there are [still a concerns brought up before](https://github.com/trinodb/trino/pull/11772#issuecomment-1191133973).

We'll try to get to the bottom of all the blockers and discuss the result of this conversation next week.

**Review by phd3 — COMMENTED** [review]
some comments, still reviewing.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/jdbc/TestIcebergJdbcConnectorTest.java
Comment (ebyhr) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/jdbc/TestIcebergJdbcConnectorTest.java:221]: Actually, I didn't investigate the details because the review process didn't start. I will look into it when I have time.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalog.java
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalog.java]: Those methods throw UncheckedSQLException when the identifier length is too long. Changed to catch RuntimeException.

**Comment by martint** [comment]
Since this is a highly requested and anticipated feature, let's move forward and get it merged. I understand the concerns about potential backward compatibility problems, but we can add some warnings to the documentation and provide the feature "as-is". Users will have to understand there's some risk that things might break if they upgrade Trino or other engines and they break compatibility.

**Comment by martint** [comment]
cc @mosabua @bitsondatadev for how to best provide those warnings in the docs.

**Comment by bitsondatadev** [comment]
@ebyhr, let's discuss some phrasing on Slack.

**Comment by ebyhr** [comment]
Rebased on upstream to resolve conflicts.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/jdbc/TestIcebergJdbcConnectorTest.java
Comment (ebyhr) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/jdbc/TestIcebergJdbcConnectorTest.java]: Just fixed the current issue.

**Comment by ebyhr** [comment]
Addressed comments.

**Review by findepi — APPROVED** [review]
Please give @alexjo2144 or @findinpath a chance to read too

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcClient.java
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcClient.java:27]: `catalogName` seems more fitting for a String value to me

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcConnectionFactory.java
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcConnectionFactory.java:39]: Should we have some connection pool here to draw from?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalog.java
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalog.java:131]: Grammar nit pick, looks like a copy from the REST wording
```suggestion
        throw new TrinoException(NOT_SUPPORTED, "setNamespacePrincipal is not supported for Iceberg JDBC catalogs");
```

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalog.java
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalog.java:147]: Catch a `NoSuchNamespaceException` from `jdbcCatalog.listTables` and ignore it. May happen that a schema is dropped between collecting the Namespaces and loading the tables from each.

**Review by alexjo2144 — COMMENTED** [review]
Are we respecting the `jdbc.strict-mode` [table property](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/jdbc/JdbcUtil.java#L37)?

Do we have follow up issues for Views and MVs?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcConnectionFactory.java
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/IcebergJdbcConnectionFactory.java:39]: I would skip supporting connection pools in this PR. We can add it later when needed.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalog.java
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/jdbc/TrinoJdbcCatalog.java]: Good catch, updated.

**Comment by ebyhr** [comment]
> Are we respecting the jdbc.strict-mode [table property](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/jdbc/JdbcUtil.java#L37)?

We shouldn't create tables on non-existing namespaces regardless of the properties and it's test by `BaseConnectorTest#testCreateTableSchemaNotFound`. The test doesn't ensure the concurrent modification though.

> Do we have follow up issues for Views and MVs?

No because there's no plan to support them at this time.

**Comment by ebyhr** [comment]
(Rebased on upstream to resolve conflicts)

**Review by alexjo2144 — APPROVED** [review]
Looks good to me

File: docs/src/main/sphinx/connector/iceberg.rst
Comment (bitsondatadev) [inline docs/src/main/sphinx/connector/iceberg.rst:127]:
```suggestion
  Consider the :ref:`REST catalog <iceberg-rest-catalog>` as an alternative solution.
```

Add 
```
.. _iceberg-rest-catalog:
```

above

```
REST catalog
^^^^^^^^^^^^^^
```

File: docs/src/main/sphinx/connector/iceberg.rst
Comment (bitsondatadev) [inline docs/src/main/sphinx/connector/iceberg.rst:131]:
```suggestion
When using any database besides PostgreSQL, a JDBC driver jar file must be placed in the plugin directory.
```

**Review by bitsondatadev — APPROVED** [review]
Just a docs review here.

Few small changes to the docs but they look good!

**Comment by bjackson-ep** [comment]
Greetings @findepi and @bitsondatadev thank you for your diligent work on JDBC other connectors.  Is there anything you need from the community to get this one across the line?  I am happy to connect you with Dremio PMs or others that can provide clarifications or support.  I see a lot of value in Trino and would love to see the query-where-it-lives grow to this catalog.  We are believers in data as code and want to help evolve that space through technologies like Trino and Nessie.  Thank you.

**Comment by alexjo2144** [comment]
> Is there anything you need from the community to get this one across the line?

The JDBC catalog was included in the 406 release a few weeks ago, so I would call it across the line. https://trino.io/docs/current/release/release-406.html#iceberg-connector

Let us know if you have a chance to try it out, feedback is always welcome.

**Comment by bjackson-ep** [comment]
https://github.com/trinodb/trino/pull/11701   Support for Nessie Catalogs is what I am talking about.  There is a March 29, 2022 pull request in the queue for this.

**Comment by penggewudi** [comment]
> 



> @sungwy The root cause relates to #17. We prefer small start when adding a new enhancement, so I won't handle case insensitive matching in this PR. We can improve such things later.

Excuse me, I would like to ask how is the progress of this issue? We connect to Iceberg through the Postgre JDBC catalog, the following is the attribute configuration of the configuration catalog.
```
connector.name=iceberg
hive.s3.endpoint=xxx
hive.s3.aws-access-key=xxx
hive.s3.aws-secret-key=xxx
hive.s3.path-style-access=true
hive.s3.ssl.enabled=false
iceberg.catalog.type=jdbc
iceberg.jdbc-catalog.catalog-name=jdbc
iceberg.jdbc-catalog.driver-class=org.postgresql.Driver
iceberg.jdbc-catalog.connection-url=jdbc:postgresql://xxx/jdbc
iceberg.jdbc-catalog.connection-user=xxx
iceberg.jdbc-catalog.connection-password=xxx
iceberg.jdbc-catalog.default-warehouse-dir=s3://xxx/
```
We use spark sql to write a table named `Test_Table` into Iceberg, When querying with the following statement through Trino CLI, the error of Table 'xxx.test_table' does not exist will appear. I try to add parameters like `iceberg.jdbc-catalog.case-insensitive-name-matching=true`, it seems that Trino service It will report an error and cannot start.

`SELECT * FROM Test_Table / SELECT * FROM test_table`

In addition, when we write to Iceberg in the same way but the table named `test_table`, I can query and get the results I want. If I want to keep the consistency of writing and reading table names, how can I do it Woolen cloth?

**Pull request merged**
